### PR TITLE
Add support for impl blocks for functions and constants.

### PR DESCRIPTION
**Pull request by philberty**
This is the building block in adding support for methods and the self keyword.
    
    Fixes: #111
    
    Fix type resolution null pointer crash
    
    This also enhance ArithmeticOrLogicalExpr rules and associated error
    handling to propagate up the stack.

### CODE REVIEW

File: gcc/rust/ast/rust-ast.h [hunk 1]
```diff
@@ -1333,6 +1333,8 @@ class InherentImplItem
 
   virtual void mark_for_strip () = 0;
   virtual bool is_marked_for_strip () const = 0;
+
+  virtual Location get_locus_slow () = 0;
```
Comment (SimplyTheOther) [inline gcc/rust/ast/rust-ast.h:1337]: Nitpick - this should be const-qualified.

File: gcc/rust/ast/rust-ast.h [hunk 2]
```diff
@@ -1459,6 +1461,10 @@ class MacroInvocationSemi : public MacroItem,
   const std::vector<Attribute> &get_outer_attrs () const { return outer_attrs; }
   std::vector<Attribute> &get_outer_attrs () { return outer_attrs; }
 
+  Location get_locus () const { return locus; }
+
+  Location get_locus_slow () { return get_locus (); }
```
Comment (SimplyTheOther) [inline gcc/rust/ast/rust-ast.h:1466]: It would probably be best to qualify this with `const final override`.

File: gcc/rust/ast/rust-item.h [hunk 3]
```diff
@@ -860,6 +867,14 @@ class Method : public InherentImplItem, public TraitImplItem
     return where_clause;
   }
 
+  Identifier get_method_name () const { return method_name; }
+
+  NodeId get_node_id () const { return node_id; }
+
+  Location get_locus () const { return locus; }
+
+  Location get_locus_slow () { return get_locus (); }
```
Comment (SimplyTheOther) [inline gcc/rust/ast/rust-item.h:876]: It would probably be best to qualify this with `const final override`.

File: gcc/rust/ast/rust-item.h [hunk 4]
```diff
@@ -1503,6 +1518,8 @@ class Function : public VisItem, public InherentImplItem, public TraitImplItem
 
   Location get_locus () const { return locus; }
 
+  Location get_locus_slow () { return get_locus (); }
```
Comment (SimplyTheOther) [inline gcc/rust/ast/rust-item.h:1521]: It would probably be best to qualify this with `const final override`.

File: gcc/rust/ast/rust-item.h [hunk 5]
```diff
@@ -2536,6 +2553,8 @@ class ConstantItem : public VisItem,
...
```
Comment (SimplyTheOther) [inline gcc/rust/ast/rust-item.h:2556]: It would probably be best to qualify this with `const final override`.

**Review by SimplyTheOther — APPROVED** [review]
PR looks great. Had some minor nitpicks regarding `get_locus_slow`.

File: gcc/rust/ast/rust-ast.h (hunk 1 above)
Comment (philberty) [reply gcc/rust/ast/rust-ast.h:1337]: Thanks for the review i will fix these before merge

**Pull request merged**